src/qbe_ir_generator.cpp

```diff
@@ -144,65 +176,104 @@ void QbeIrGenerator::Visit(const DeclStmtNode& decl_stmt) {
+      Write_("export data {0} = align {1} {{ z {1} }}\n",
+             user_defined::GlobalPointer{decl.id}, decl.type->size());
+    }
+  } else {
+    int id_num = NextLocalNum();
+    WriteInstr_("{} =l alloc{} {}", FuncScopeTemp{id_num}, decl.type->size(),
+                decl.type->size());
+    if (decl.init) {
+      decl.init->Accept(*this);
+      int init_num = num_recorder.NumOfPrevExpr();
+      // A pointer declaration may have two options for its right hand side:
+      if (decl.init->type->IsPtr() || decl.init->type->IsFunc()) {
+        // 1. int* a = &b; rhs is a reference of integer. We need to store b's
+        // address to a, where we need to map b's reg_num back to its id_num.
+        if (dynamic_cast<UnaryExprNode*>((decl.init).get())) {
+          WriteInstr_("storel {}, {}",
+                      FuncScopeTemp{reg_num_to_id_num.at(init_num)},
+                      FuncScopeTemp{id_num});
+        } else {
+          // 2. int* a = c; c itself stores the address of another integer. We
+          // can directly use the address c currently holds.
+          WriteInstr_("storel {}, {}", FuncScopeTemp{init_num},
+                      FuncScopeTemp{id_num});
+        }
       } else {
-        // 2. int* a = c; c itself stores the address of another integer. We can
-        // directly use the address c currently holds.
-        WriteInstr_("storel {}, {}", FuncScopeTemp{init_num},
+        WriteInstr_("storew {}, {}", FuncScopeTemp{init_num},
                     FuncScopeTemp{id_num});
       }
-    } else {
-      WriteInstr_("storew {}, {}", FuncScopeTemp{init_num},
-                  FuncScopeTemp{id_num});
     }
+    // Set up the number of the id so we know were to load it back.
+    id_to_num[decl.id] = id_num;
   }
-  // Set up the number of the id so we know were to load it back.
-  id_to_num[decl.id] = id_num;
 }
 
 void QbeIrGenerator::Visit(const ArrDeclNode& arr_decl) {
-  int base_addr_num = NextLocalNum();
-  assert(arr_decl.type->IsArr());
-  const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
-  auto element_size = arr_type->element_type().size();
-  WriteInstr_("{} =l alloc{} {}", FuncScopeTemp{base_addr_num}, element_size,
-              arr_decl.type->size());
-  id_to_num[arr_decl.id] = base_addr_num;
-
-  for (auto i = std::size_t{0}, e = arr_type->len(); i < e; ++i) {
-    if (i < arr_decl.init_list.size()) {
-      auto& arr_init = arr_decl.init_list.at(i);
-      arr_init->Accept(*this);
+  if (arr_decl.is_global) {
+    const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
+    assert(arr_type);
+    if (arr_decl.init_list.size() != 0) {
+      global_var_init_vals.clear();
+      Write_("export data {} = align {} {{ ",
+             user_defined::GlobalPointer{arr_decl.id},
+             arr_type->element_type().size());
+      for (auto i = std::size_t{0}, e = arr_type->len(); i < e; ++i) {
+        auto& arr_init = arr_decl.init_list.at(i);
```

Comment:
Is it possible that the `init_list` is shorter than the `arr_type->len()`? If yes, those elements have to be set as 0, which can be done using `z {remaining bytes}`.

Reply:
Updated!